Panic during wallet file-watcher notification

# Problem encountered
```
panic: runtime error: slice bounds out of range [:12] with capacity 1

goroutine 116 [running]:
code.vegaprotocol.io/vega/libs/crypto.Decrypt({0xc001886079, 0x0, 0x1}, {0xc000fa4580?, 0x181cff1?})
	/jenkins/workspace/common/system-tests-wrapper/vega/libs/crypto/encryption.go:35 +0x165
code.vegaprotocol.io/vega/wallet/wallet/store/v1.(*FileStore).readWalletFile(0xc001d1a690, {0xc000dcf720, 0xa}, {0xc000fa4580, 0x1c})
	/jenkins/workspace/common/system-tests-wrapper/vega/wallet/wallet/store/v1/file_store.go:390 +0x13f
code.vegaprotocol.io/vega/wallet/wallet/store/v1.(*FileStore).handleWalletFileUpdate(0xc001d1a690, {0x52e5198, 0xc001c01500}, {0xc000f20057, 0xa})
	/jenkins/workspace/common/system-tests-wrapper/vega/wallet/wallet/store/v1/file_store.go:532 +0xa5
code.vegaprotocol.io/vega/wallet/wallet/store/v1.(*FileStore).applyFileChangesToStore(0xc001d1a690, {0x52e5198, 0xc001c01500}, {{0xc000f20000?, 0x0?}, 0x1?})
	/jenkins/workspace/common/system-tests-wrapper/vega/wallet/wallet/store/v1/file_store.go:521 +0x185
code.vegaprotocol.io/vega/wallet/wallet/store/v1.(*FileStore).watchFile(0xc001d1a690, {0x52e5198, 0xc001c01500})
	/jenkins/workspace/common/system-tests-wrapper/vega/wallet/wallet/store/v1/file_store.go:468 +0x156
code.vegaprotocol.io/vega/wallet/wallet/store/v1.(*FileStore).startFilesWatcher.func1({0x52e5198?, 0xc001c01500?})
	/jenkins/workspace/common/system-tests-wrapper/vega/wallet/wallet/store/v1/file_store.go:445 +0x30
code.vegaprotocol.io/vega/libs/job.(*Runner).Go.func1()
	/jenkins/workspace/common/system-tests-wrapper/vega/libs/job/runner.go:31 +0x66
created by code.vegaprotocol.io/vega/libs/job.(*Runner).Go
	/jenkins/workspace/common/system-tests-wrapper/vega/libs/job/runner.go:29 +0x8a
```

The system-tests are in the process of being wallet-V2 enabled. While running a single test the wallet panicked while setting up the network, the following CLI commands were called while the service was running:
```
# a wallet is imported via the CLI
vega wallet import --output json --wallet 19C11579E9 --passphrase-file PASSPHRASE_FILE --recovery-phrase-file RECOVERY_FILE

# A long lived API token is created for the wallet
vega wallet api-token generate  --output json --wallet-name 19C11579E9 --wallet-passphrase-file PASSPHRASE_FILE --tokens-passphrase-file PASSPHRASE_FILE --description 19C11579E9

# we attempt to list the keys through the V2 API
data = {
            "jsonrpc": "2.0",
            "method": "client.list_keys",
            "id": "1",
        }
r = self._request(data) # fails because the wallet is dead
```

I don't know where in that chain the file watcher kicked in. Judging by the panic it looks like when the file was read it was empty? Possibly some race-condition somewhere.

I imagine this will be hard to reproduce since I have done many full runs over the past few days and this is the first time I've encountered this.

Jenkins run: https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests-wrapper/detail/system-tests-wrapper/48353/pipeline

Wallet logs: https://jenkins.ops.vega.xyz/job/common/job/system-tests-wrapper/48353/artifact/testnet/logs/testnet-vegawallet/

# Observed behaviour
Wallet dies.

# Expected behaviour
Wallet does not die.

# System response
Describe what the system response was, include the output from the command, automation, or else.

# Steps to reproduce

## Manual
Steps to reproduce the behaviour manually:
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

## Automation
Link to automation and explanation on how to run it to reproduce the problem/bug

# Evidence

## Logs
If applicable, add logs and/or screenshots to help explain your problem.

## Additional context
Add any other context about the problem here including; system version numbers, components affected.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -34,6 +34,7 @@
 - [7407](https://github.com/vegaprotocol/vega/issues/7407) - fix `ethereum` timestamp in stake linking in `graphql`
 - [7420](https://github.com/vegaprotocol/vega/issues/7420) - `clearFeeActivity` now clears fee activity
 - [7420](https://github.com/vegaprotocol/vega/issues/7420) - set seed nonce for joining and leaving signatures during begin block
+- [7420](https://github.com/vegaprotocol/vega/issues/7515) - protect `vegawallet` with recovers to shield against file system oddities
 - [7399](https://github.com/vegaprotocol/vega/issues/7399) - Fix issue where market cache not working after restoring from network history
 - [7410](https://github.com/vegaprotocol/vega/issues/7410) - Return underlying error when parsing a command failed in the wallet API version 2
 - [7169](https://github.com/vegaprotocol/vega/issues/7169) - Fix migration, account for existing position data

--- a/wallet/service/v2/connections/store/v1/file_store.go
+++ b/wallet/service/v2/connections/store/v1/file_store.go
@@ -173,7 +173,7 @@ func (s *FileStore) Close() {
 func (s *FileStore) readTokensFile() (tokens tokensFile, rerr error) {
 	defer func() {
 		if r := recover(); r != nil {
-			tokens, rerr = tokensFile{}, fmt.Errorf("unexpected fs issues reading tokens file: %s", r)
+			tokens, rerr = tokensFile{}, fmt.Errorf("a system error occurred while reading the tokens file: %s", r)
 		}
 	}()
 
@@ -205,7 +205,7 @@ func (s *FileStore) readTokensFile() (tokens tokensFile, rerr error) {
 func (s *FileStore) writeTokensFile(tokens tokensFile) (rerr error) {
 	defer func() {
 		if r := recover(); r != nil {
-			rerr = fmt.Errorf("unexpected fs issues writing tokens file: %s", r)
+			rerr = fmt.Errorf("a system error occurred while writing the tokens file:: %s", r)
 		}
 	}()
 	if err := paths.WriteEncryptedFile(s.tokensFilePath, s.passphrase, tokens); err != nil {

--- a/wallet/service/v2/connections/store/v1/file_store.go
+++ b/wallet/service/v2/connections/store/v1/file_store.go
@@ -170,8 +170,12 @@ func (s *FileStore) Close() {
 	}
 }
 
-func (s *FileStore) readTokensFile() (tokensFile, error) {
-	tokens := tokensFile{}
+func (s *FileStore) readTokensFile() (tokens tokensFile, rerr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			tokens, rerr = tokensFile{}, fmt.Errorf("unexpected fs issues reading tokens file: %s", r)
+		}
+	}()
 
 	exists, err := vgfs.FileExists(s.tokensFilePath)
 	if err != nil {
@@ -198,7 +202,12 @@ func (s *FileStore) readTokensFile() (tokensFile, error) {
 	return tokens, nil
 }
 
-func (s *FileStore) writeTokensFile(tokens tokensFile) error {
+func (s *FileStore) writeTokensFile(tokens tokensFile) (rerr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			rerr = fmt.Errorf("unexpected fs issues writing tokens file: %s", r)
+		}
+	}()
 	if err := paths.WriteEncryptedFile(s.tokensFilePath, s.passphrase, tokens); err != nil {
 		return fmt.Errorf("couldn't write the file %s: %w", s.tokensFilePath, err)
 	}

--- a/wallet/wallet/store/v1/file_store.go
+++ b/wallet/wallet/store/v1/file_store.go
@@ -23,6 +23,7 @@ import (
 var (
 	ErrWalletNameCannotContainSlashCharacters = errors.New("the name cannot contain slash (\"/\", \"\\\") characters")
 	ErrWalletNameCannotStartWithDot           = errors.New("the name cannot start with a dot (\".\") character")
+	ErrWalletFileIsEmpty                      = errors.New("the wallet file is empty")
 )
 
 type FileStore struct {
@@ -353,7 +354,12 @@ func (s *FileStore) GetWalletPath(name string) string {
 	return s.walletPath(name)
 }
 
-func (s *FileStore) writeWallet(w wallet.Wallet, passphrase string) error {
+func (s *FileStore) writeWallet(w wallet.Wallet, passphrase string) (rerr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			rerr = fmt.Errorf("unexpected fs issues writing wallet file: %s", r)
+		}
+	}()
 	if err := ensureValidWalletName(w.Name()); err != nil {
 		return err
 	}
@@ -381,10 +387,20 @@ func (s *FileStore) walletPath(name string) string {
 	return filepath.Join(s.walletsHome, name)
 }
 
-func (s *FileStore) readWalletFile(name string, passphrase string) (*wallet.HDWallet, error) {
+func (s *FileStore) readWalletFile(name string, passphrase string) (hd *wallet.HDWallet, rerr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			hd, rerr = nil, fmt.Errorf("unexpected fs issues reading wallet file: %s", r)
+		}
+	}()
+
 	buf, err := fs.ReadFile(os.DirFS(s.walletsHome), name)
 	if err != nil {
 		return nil, fmt.Errorf("could not read the wallet file at %s: %w", s.walletsHome, err)
+	}
+
+	if len(buf) == 0 {
+		return nil, ErrWalletFileIsEmpty
 	}
 
 	decBuf, err := vgcrypto.Decrypt(buf, passphrase)

--- a/wallet/wallet/store/v1/file_store.go
+++ b/wallet/wallet/store/v1/file_store.go
@@ -357,7 +357,7 @@ func (s *FileStore) GetWalletPath(name string) string {
 func (s *FileStore) writeWallet(w wallet.Wallet, passphrase string) (rerr error) {
 	defer func() {
 		if r := recover(); r != nil {
-			rerr = fmt.Errorf("unexpected fs issues writing wallet file: %s", r)
+			rerr = fmt.Errorf("a system error occurred while writing the wallet file: %s", r)
 		}
 	}()
 	if err := ensureValidWalletName(w.Name()); err != nil {
@@ -390,7 +390,7 @@ func (s *FileStore) walletPath(name string) string {
 func (s *FileStore) readWalletFile(name string, passphrase string) (hd *wallet.HDWallet, rerr error) {
 	defer func() {
 		if r := recover(); r != nil {
-			hd, rerr = nil, fmt.Errorf("unexpected fs issues reading wallet file: %s", r)
+			hd, rerr = nil, fmt.Errorf("a system error occurred while reading the wallet file: %s", r)
 		}
 	}()
 

